Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,3028 @@
+    if (prevObstacle.isValid()) {
+        // The previous point is the final contact point of the previous curve.
+        return getCable()
+            .getObstacleCurveSegment(prevObstacle)
+            .calcFinalContactPoint_G(state);
+    }
+    // There are no curve segments before given obstacle: the previous point
+    // is the path's origin point.
+    const CableSpan::Impl& cable = getCable();
+    return getCable()
+        .getOriginBody()
+        .getBodyTransform(state)
+        .shiftFrameStationToBase(cable.getOriginPoint_B());
+}
+
+Vec3 CurveSegment::findNextPathPoint_G(const State& state) const
+{
+    // Check if there is a curve segment after given obstacle.
+    const ObstacleIndex nextObstacle =
+        findNextObstacleInContactWithCable(state);
+    if (nextObstacle.isValid()) {
+        // The next point is the initial contact point of the next curve.
+        return getCable()
+            .getObstacleCurveSegment(nextObstacle)
+            .calcInitialContactPoint_G(state);
+    };
+    // There are no curve segments following given obstacle: the next point
+    // is the path's termination point.
+    const CableSpan::Impl& cable = getCable();
+    return cable.getTerminationBody()
+        .getBodyTransform(state)
+        .shiftFrameStationToBase(cable.getTerminationPoint_B());
+}
+
+//==============================================================================
+//                      CABLE FORCE COMPUTATIONS
+//==============================================================================
+// This section contains all force related computations.
+namespace
+{
+
+void calcUnitForceExertedByCurve(
+    const CurveSegment& curve,
+    const State& s,
+    SpatialVec& unitForce_G)
+{
+    const CurveSegmentData::Pos& dataPos = curve.getDataPos(s);
+    const Vec3& x_GB = curve.getMobilizedBody().getBodyOriginLocation(s);
+
+    // Contact point moment arms in ground.
+    const Vec3 r_P = dataPos.X_GP.p() - x_GB;
+    const Vec3 r_Q = dataPos.X_GQ.p() - x_GB;
+
+    // Tangent directions at contact points in ground.
+    const UnitVec3& t_P = getTangent(dataPos.X_GP);
+    const UnitVec3& t_Q = getTangent(dataPos.X_GQ);
+
+    unitForce_G[0] = r_Q % t_Q - r_P % t_P;
+    unitForce_G[1] = t_Q - t_P;
+}
+
+void calcUnitForceAtCableOrigin(
+    const CableSpan::Impl& cable,
+    const State& s,
+    SpatialVec& unitForce_G)
+{
+    // Origin contact point moment arm in ground.
+    const CableSpanData::Pos& dataPos = cable.getDataPos(s);
+
+    const Vec3& arm_G =
+        dataPos.originPoint_G - cable.getOriginBody().getBodyOriginLocation(s);
+
+    unitForce_G[0] = -arm_G % dataPos.originTangent_G;
+    unitForce_G[1] = -Vec3(dataPos.originTangent_G);
+}
+
+void calcUnitForceAtCableTermination(
+    const CableSpan::Impl& cable,
+    const State& s,
+    SpatialVec& unitForce_G)
+{
+    const CableSpanData::Pos& dataPos = cable.getDataPos(s);
+
+    const Vec3& arm_G = dataPos.terminationPoint_G -
```

Comment:
I mean, you're luck that reference-lifetime-extension is a feature of C++ for const-qualified lvalues or you'd be in trouble

Reply:
huh. so that actually compiles :)